python_modules/dagster/dagster/_core/instance/__init__.py

```diff
@@ -1510,6 +1506,26 @@ def get_run():
         except DagsterRunAlreadyExists:
             return get_run()
 
+    def handle_started_run(self, run_record: RunRecord, logger: logging.Logger) -> None:
+        max_time_str = run_record.dagster_run.tags.get(MAX_RUNTIME_TAG)
+
+        # If MAX_RUNTIME_TAG has not been set, nothing to do
+        if not max_time_str:
+            return
+        max_time = float(max_time_str)
+        if (
+            run_record.start_time is not None
+            and pendulum.now().timestamp() - run_record.start_time > max_time
+        ):
+            logger.info(
+                f"Run {run_record.dagster_run.run_id} has exceeded maximum runtime of"
+                f" {max_time} seconds: terminating run."
+            )
+            self.run_launcher.terminate(
+                run_id=run_record.dagster_run.run_id,
+                message="Run has exceeded maximum allowed runtime.",
+            )
```

Comment:
i think we may want something like this where it moves it into a CANCELED state no matter what but still tries to clean it up (instead of maybe possibly terminating it if everything works out): https://github.com/dagster-io/dagster/blob/master/python_modules/dagster-graphql/dagster_graphql/implementation/execution/__init__.py#L104-L121

Arguably all run terminations should work this way